compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/DecomposeLinalgExtPackUnPackToAIR.cpp

```diff
@@ -0,0 +1,505 @@
+    SmallVector<OpFoldResult> readStrides(srcRank, oneIdxAttr);
+    SmallVector<OpFoldResult> readSizes;
+    SmallVector<int64_t> readShape;
+    for (auto i : llvm::seq<unsigned>(0, srcRank)) {
+      if (dimAndTileMapping.count(i)) {
+        readShape.push_back(getConstantIntValue(dimAndTileMapping[i])
+                                .value_or(ShapedType::kDynamic));
+        readSizes.push_back(dimAndTileMapping[i]);
+        continue;
+      }
+      if (ShapedType::isDynamic(inputShape[i])) {
+        assert(false);
+        // TODO: Dynamic input shape
+      } else {
+        readSizes.push_back(rewriter.getIndexAttr(inputShape[i]));
+      }
+      readShape.push_back(inputShape[i]);
+    }
+
+    Type elemType = packOp.getInputType().getElementType();
+    Attribute memorySpace =
+        packOp.getInputType().cast<MemRefType>().getMemorySpace();
+    auto readType = MemRefType::get(
+        readShape, elemType,
+        packOp.getInputType().cast<MemRefType>().getLayout(), memorySpace);
+
+    Value tile = rewriter.create<memref::SubViewOp>(
+        loc, readType, input, readOffsets, readSizes, readStrides);
+
+    // 2. Transpose the tile to match the inner tile order.
+
+    SmallVector<int64_t> perm = getPackUnpackRankReducedPerm(
+        inputShape, innerDimsPos, packOp.getOuterDimsPerm());
+
+    SmallVector<int64_t> transpShape = readShape;
+    applyPermutationToVector<int64_t>(transpShape, perm);
+
+    Value transposed = rewriter.create<memref::TransposeOp>(
+        loc, tile,
+        AffineMapAttr::get(
+            AffineMap::getPermutationMap(perm, packOp->getContext())));
+
+    // 3. Insert the inner tile to the destination.
+    int64_t destRank = packOp.getOutputRank();
+    SmallVector<OpFoldResult> writeStrides(destRank, oneIdxAttr);
+    SmallVector<OpFoldResult> writeOffsets(destRank, zeroIdxAttr);
+    SmallVector<OpFoldResult> writeSizes =
+        memref::getMixedSizes(rewriter, loc, packOp.getOutput());
+
+    memorySpace = packOp.getOutputType().cast<MemRefType>().getMemorySpace();
+    auto writeType = MemRefType::get(destShape, elemType, nullptr, memorySpace);
+    auto output_subview = rewriter.create<memref::SubViewOp>(
+        loc, writeType, packOp.getOutput(), writeOffsets, writeSizes,
+        writeStrides);
+
+    SmallVector<Value, 2> emptyVec;
+    rewriter.create<xilinx::air::DmaMemcpyNdOp>(
+        loc, SmallVector<Type, 1>{}, emptyVec, output_subview, emptyVec,
+        emptyVec, emptyVec, transposed, emptyVec, emptyVec, emptyVec);
+
+    rewriter.eraseOp(packOp);
+    return success();
+  }
+};
+
+class GeneralizeUnPackOpPattern
+    : public OpRewritePattern<IREE::LinalgExt::UnPackOp> {
+ public:
+  using OpRewritePattern<IREE::LinalgExt::UnPackOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(IREE::LinalgExt::UnPackOp unpackOp,
+                                PatternRewriter &rewriter) const override {
+    int64_t srcRank = unpackOp.getInputRank();
```

Comment:
I think you can common out a lot of this implementation by moving the core logic to a common function. `unpack` and `pack` are just reverse of each other, so just flipping the shapes of the input/output should work out the same way.

Reply:
Agreed. Moved inference of transform perms to a common function.